docs/specs/app2app.md

```diff
@@ -21,20 +21,38 @@ An app can start the authentication flow by opening a link to another app, inste
 
 ### Server-side
 
-- Authorization Endpoint
+- Configuration
 
-  - A new parameter `device_key` is supported. This value can be specified by the client to enable app2app login. If specified, it should be a public key of a key pair, which the private key should be stored in a secure storage in the device. Such as the Android keystore or the iOS keychain.
-    - Multiple key types such as RS256 and NIST P-256 are supported.
+  - A app2app section will be added to the client configs.
+
+  ```yaml
+  oauth:
+    clients:
+      - client_id: CLIENT_ID
+        # ...Other configs
+        x_app2app:
+          enabled: true
+          allow_insecure_device_key_binding: false
+  ```
+
+  - `enabled`: boolean. Whether the client is able to handle app2app authentication requests from other clients.
+  - `allow_insecure_device_key_binding`: boolean. Default `false`. If `true`, refresh tokens of this client without a binded `device_key` can be binded to a new `device_key` during refreshing the tokens. This option is for allowing existing logged in users to use app2app without requiring them to re-login. However this also allows sessions without a binded `device_key` to be binded with any key and participate in app2app authentication which might be a security concern.
 
 - Token Endpoint
 
+  - For `grant_type=authorization_code`, a new parameter `device_key` is supported. This value can be specified by the client to enable app2app login. If specified, it should be a base64 encoded jwk of public key of a key pair, which the private key should be stored in a secure storage in the device. Such as the Android keystore or the iOS keychain. `device_key` will be ignored if `x_app2app.enabled` of the client configuration is not `true`.
```

Comment:
Rename this to `x_app2app_device_key_jwt` to signify it is our own extension.

And we are going to make it a JWT.

header
```
{
  "alg": "ES256",
  "kty": "EC",
  "kid": "randomly-generated-kid",
  "jwk": { ... }
}
```

Payload
```
{
  "iat": 123,
  "exp": 123,
   "challenge": "one-time-use-challenge-obtain-from-server"
}
```

Reply:
@louischan-oursky Why making it a jwt? This is the public key to bind to the refresh token but not to verify the user owns the private key.

Reply:
1. It is a proof that the caller has the private key. We will verify the signature of the JWT with the JWK inside it.
2. Avoid replay attack. We will check if the challenge is consumed or not.